File: packages/php/src/Services/AccountService.php
<?php

namespace Yourdudeken\Mpesa\Services;

use Yourdudeken\Mpesa\Http\Client;
use Yourdudeken\Mpesa\Helpers\Signature;

class AccountService
{
    private $client;
    private $signature;

    public function __construct(Client $client, Signature $signature)
    {
        $this->client = $client;
        $this->signature = $signature;
    }

    public function balance($shortcode, $identifierType, $remarks, $result_url = null, $timeout_url = null, $shortCodeType = 'C2B')
    {
        $url = $this->client->getBaseUrl() . '/mpesa/accountbalance/v1/query';

        $body = [
            'Initiator' => config('mpesa.initiator_name'),
            'SecurityCredential' => $this->signature->generateSecurityCredential(),
            'CommandID' => 'AccountBalance',
            'PartyA' => $shortcode,
            'IdentifierType' => $identifierType,
            'Remarks' => $remarks,
            'ResultURL' => $this->resolveCallbackUrl($result_url, 'balance_result_url'),
            'QueueTimeOutURL' => $this->resolveCallbackUrl($timeout_url, 'balance_timeout_url'),
        ];

        return $this->client->post($url, $body, $shortCodeType);
    }

    public function status($shortcode, $transactionId, $identifierType, $remarks, $result_url = null, $timeout_url = null, $shortCodeType = 'C2B')
    {
        $url = $this->client->getBaseUrl() . '/mpesa/transactionstatus/v1/query';

        $body = [
            'Initiator' => config('mpesa.initiator_name'),
            'SecurityCredential' => $this->signature->generateSecurityCredential(),
            'CommandID' => 'TransactionStatusQuery',
            'TransactionID' => $transactionId,
            'PartyA' => $shortcode,
            'IdentifierType' => $identifierType,
            'Remarks' => $remarks,
            'Occassion' => '',
            'ResultURL' => $this->resolveCallbackUrl($result_url, 'status_result_url'),
            'QueueTimeOutURL' => $this->resolveCallbackUrl($timeout_url, 'status_timeout_url'),
        ];

        return $this->client->post($url, $body, $shortCodeType);
    }

    public function reversal($shortcode, $transactionId, $amount, $remarks, $result_url = null, $timeout_url = null, $shortCodeType = 'C2B')
    {
        $url = $this->client->getBaseUrl() . '/mpesa/reversal/v1/request';

        $body = [
            'Initiator' => config('mpesa.initiator_name'),
            'SecurityCredential' => $this->signature->generateSecurityCredential(),
            'CommandID' => 'TransactionReversal',
            'TransactionID' => $transactionId,
            'Amount' => $amount,
            'ReceiverParty' => $shortcode,
            'RecieverIdentifierType' => '11',
            'Remarks' => $remarks,
            'Occasion' => '',
            'ResultURL' => $this->resolveCallbackUrl($result_url, 'reversal_result_url'),
            'QueueTimeOutURL' => $this->resolveCallbackUrl($timeout_url, 'reversal_timeout_url'),
        ];

        return $this->client->post($url, $body, $shortCodeType);
    }

    private function resolveCallbackUrl($paramUrl, $configKey)
    {
        $configUrl = config("mpesa.callbacks.{$configKey}");
        
        if ($paramUrl !== null) {
            return $paramUrl;
        } elseif ($configUrl !== null) {
            return $configUrl;
        }
        
        throw new \Exception("Ensure you have set the {$configKey} in the config or passed as a parameter");
    }
}

File: packages/go/mpesa/account_service.go
package mpesa

type AccountService struct {
	httpClient *HTTPClient
	auth       *Auth
	helpers    *Helpers
	baseURL    string
}

func NewAccountService(httpClient *HTTPClient, auth *Auth, config *Config) *AccountService {
	return &AccountService{
		httpClient: httpClient,
		auth:       auth,
		helpers:    NewHelpers(config),
		baseURL:    httpClient.GetBaseURL(),
	}
}

func (s *AccountService) Balance(shortcode string, identifierType int, remarks string, resultURL string, timeoutURL string, shortCodeType string) (map[string]interface{}, error) {
	url := s.baseURL + "/mpesa/accountbalance/v1/query"
	body := map[string]interface{}{
		"Initiator":          s.helpers.GetConfig("initiator_name"),
		"SecurityCredential": "",
		"CommandID":          "AccountBalance",
		"PartyA":             shortcode,
		"IdentifierType":     identifierType,
		"Remarks":            remarks,
		"ResultURL":          resultURL,
		"QueueTimeOutURL":    timeoutURL,
	}

	token, err := s.auth.GetAccessToken(shortCodeType)
	if err != nil {
		return nil, err
	}

	return s.httpClient.Post(url, body, token)
}

func (s *AccountService) Status(shortcode string, transactionID string, identifierType int, remarks string, resultURL string, timeoutURL string, shortCodeType string) (map[string]interface{}, error) {
	url := s.baseURL + "/mpesa/transactionstatus/v1/query"
	body := map[string]interface{}{
		"Initiator":          s.helpers.GetConfig("initiator_name"),
		"SecurityCredential": "",
		"CommandID":          "TransactionStatusQuery",
		"TransactionID":      transactionID,
		"PartyA":             shortcode,
		"IdentifierType":     identifierType,
		"Remarks":            remarks,
		"Occassion":          "",
		"ResultURL":          resultURL,
		"QueueTimeOutURL":    timeoutURL,
	}

	token, err := s.auth.GetAccessToken(shortCodeType)
	if err != nil {
		return nil, err
	}

	return s.httpClient.Post(url, body, token)
}

func (s *AccountService) Reversal(shortcode string, transactionID string, amount float64, remarks string, resultURL string, timeoutURL string, shortCodeType string) (map[string]interface{}, error) {
	url := s.baseURL + "/mpesa/reversal/v1/request"
	body := map[string]interface{}{
		"Initiator":              s.helpers.GetConfig("initiator_name"),
		"SecurityCredential":     "",
		"CommandID":              "TransactionReversal",
		"TransactionID":          transactionID,
		"Amount":                 amount,
		"ReceiverParty":          shortcode,
		"RecieverIdentifierType": "11",
		"Remarks":                remarks,
		"Occasion":               "",
		"ResultURL":              resultURL,
		"QueueTimeOutURL":        timeoutURL,
	}

	token, err := s.auth.GetAccessToken(shortCodeType)
	if err != nil {
		return nil, err
	}

	return s.httpClient.Post(url, body, token)
}
